### PR TITLE
tree.repo: fix race condition when walking subrepos

### PR DESCRIPTION
**Pull request by skshetry**
The subrepos were not removed from getting traversed, due to which it was causing race conditions, as we were expecting
different directories to be traversed but as subrepos were still there, `next(repo_walk)` could consume from the subrepo instead of the expected directories.

This was too hard to debug :disappointed_relieved:.

* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

### CODE REVIEW

File: tests/unit/tree/test_repo.py [hunk 1]
```diff
@@ -398,3 +398,54 @@ def test_get_hash_cached_granular(tmp_dir, dvc, mocker):
         "md5", "8d777f385d3dfec8815d20f7496026dc",
     )
     assert dvc_tree_spy.called
+
+
+@pytest.mark.parametrize("traverse_subrepos", [True, False])
+def test_walk_nested_subrepos(tmp_dir, dvc, scm, traverse_subrepos):
```
Comment (skshetry) [inline tests/unit/tree/test_repo.py:404]: The tests setup is a bit lengthy.

**Comment by skshetry** [comment]
To clarify: the bug might affect any directory that contains at least a sub-repo and other files/dirs when `walk`-ing.

File: tests/unit/tree/test_repo.py (hunk 1 above)
Comment (skshetry) [reply tests/unit/tree/test_repo.py:404]: Could not reduce it. Tell me if it's confusing.

**Review by efiop — APPROVED** [review]

**Comment by efiop** [comment]
Good catch!

**Pull request merged**